internal/provider/resource_apply_module.go

```diff
@@ -0,0 +1,596 @@
+	}
+	if finishedRun.ModuleVersion == nil {
+		diags.AddError("Finished run's module version is nil.", "")
+		return diags
+	}
+
+	// Return the output.
+	output.WorkspacePath = types.StringValue(finishedRun.WorkspacePath)
+	output.ModuleSource = types.StringValue(*finishedRun.ModuleSource)
+	output.ModuleVersion = types.StringValue(*finishedRun.ModuleVersion)
+	output.Variables = input.model.Variables // Cannot get variables back from a workspace or run, so pass them through.
+	return nil
+}
+
+func (t *applyModuleResource) waitForJobCompletion(ctx context.Context, jobID *string) error {
+	if jobID == nil {
+		return fmt.Errorf("nil job ID")
+	}
+
+	// Poll until job has finished.
+	for {
+
+		job, err := t.client.Job.GetJob(ctx, &sdktypes.GetJobInput{
+			ID: *jobID,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get job ID %s", *jobID)
+		}
+
+		if job.Status == "finished" {
+			return nil
+		}
+
+		time.Sleep(jobCompletionPollInterval)
+	}
+
+}
+
+// getCurrentApplied returns an ApplyModuleModel reflecting what is currently applied.
+func (t *applyModuleResource) getCurrentApplied(ctx context.Context,
+	tfState ApplyModuleModel, target *ApplyModuleModel) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	// Get latest run on the target workspace.
+	wsPath := tfState.WorkspacePath.ValueString()
+	ws, err := t.client.Workspaces.GetWorkspace(ctx, &sdktypes.GetWorkspaceInput{
+		Path: &wsPath,
+	})
+	if err != nil {
+		diags.AddError(fmt.Sprintf("Failed to get specified workspace by path: %s", wsPath), err.Error())
+		return diags
+	}
+	latestRun, err := t.client.Run.GetRun(ctx, &sdktypes.GetRunInput{
+		ID: ws.CurrentStateVersion.RunID,
+	})
+	if err != nil {
+		diags.AddError("Failed to get latest run", err.Error())
+		return diags
+	}
+
+	// Make sure the module source and module version are not nil.
+	if latestRun.ModuleSource == nil {
+		diags.AddError("No module source available", fmt.Sprintf("for workspace %s", latestRun.WorkspacePath))
+		return diags
+	}
+	if latestRun.ModuleVersion == nil {
+		diags.AddError("No module version available", fmt.Sprintf("for workspace %s", latestRun.WorkspacePath))
+		return diags
+	}
+
+	target.WorkspacePath = tfState.WorkspacePath
+	target.ModuleSource = types.StringValue(*latestRun.ModuleSource)
+	target.ModuleVersion = types.StringValue(*latestRun.ModuleVersion)
+	target.Variables = tfState.Variables
+
+	return nil
+}
+
+// copyApplyModule copies the contents of an apply module.
+// It copies the fields from the same type, because there is not an apply module defined by Tharsis.
+func (t *applyModuleResource) copyApplyModule(ctx context.Context, src, dest *ApplyModuleModel) diag.Diagnostics {
+	dest.WorkspacePath = src.WorkspacePath
+	dest.ModuleSource = src.ModuleSource
+	dest.ModuleVersion = src.ModuleVersion
+	dest.Variables = src.Variables
+
+	// Make sure variables aren't unknown, because Terraform doesn't like that.
+	var listDiags diag.Diagnostics
+	if dest.Variables.IsUnknown() {
```

Comment:
Variables shouldn't be unknown since it's using an empty list as the default value.

Reply:
The list is marked as optional, which means it has to also be marked as computed.  If the list is omitted, it comes in to the provider as unknown.  This code is what changes that unknown to an empty list.  I had to add it during debug after Terraform errored out (saying it was a bug in the provider) on that specific case.  Or, would you prefer a different solution?

Reply:
I could be wrong but it doesn't seem like the value should be unknown since the plan modifier should be setting it to a an empty list.

Reply:
I had already done that experiment, but I re-did it to document the results.  I added this code:
![image](https://user-images.githubusercontent.com/86491630/232085003-07892993-734d-457a-81ed-3ae983cc2298.png)

The result is here: https://tharsis.dev.cts.infor.com/groups/rmr/rmr1/-/runs/Ul8wZTliNDY0My1jNmYxLTQ2MTMtYjdiNS03NTZmY2VlNDFlOTY/apply

Please note in line 45 (toward the right of the third line in my view of it) says ...IsUnknown() returns true.  Also, Terraform errors out the run, saying there's a bug in the provider.
![image](https://user-images.githubusercontent.com/86491630/232085672-e3f4eb63-8c0c-400b-9e6d-9fb2c2787dc9.png)